File: cses/leetcode/cpp/938.cpp

#include <bits/stdc++.h>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    void dfs(TreeNode *root, int *sum, int low, int high)
    {
        if (root == nullptr)
            return;
        if (root->val >= low && root->val <= high)
            *sum += root->val;
        if (root->left != nullptr)
            dfs(root->left, sum, low, high);
        if (root->right != nullptr)
            dfs(root->right, sum, low, high);
    }
    int rangeSumBST(TreeNode *root, int low, int high)
    {
        int ans = 0;
        dfs(root, &ans, low, high);
        return ans;
    }
};
int main()
{
}

File: cses/leetcode/python/938.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def rangeSumBST1(self, root: TreeNode, low: int, high: int) -> int:
        total_sum = 0
        def dfs(root):
            nonlocal total_sum
            if root is None:
                return 
            if root.val >= low and root.val <= high:
                total_sum += root.val
            dfs(root.left)
            dfs(root.right)
        dfs(root)
        return total_sum
    def rangeSumBST(self, root: TreeNode, low: int, high: int) -> int:
        if not root:
            return 0
        elif root.val < low:
            return self.rangeSumBST(root.right,low,high)
        elif root.val > high:
            return self.rangeSumBST(root.left, low, high)
        else:
            return root.val + self.rangeSumBST(root.left) + self.rangeSumBST(root.right)

        